pinot-spi/src/main/java/org/apache/pinot/spi/accounting/ThreadResourceUsageProvider.java

```diff
@@ -0,0 +1,145 @@
+ * and allocate bytes for the current thread.
+ */
+public class ThreadResourceUsageProvider {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ThreadResourceUsageProvider.class);
+  private static final String SUN_THREAD_MXBEAN_CLASS_NAME = "com.sun.management.ThreadMXBean";
+  private static final String SUN_THREAD_MXBEAN_IS_THREAD_ALLOCATED_MEMORY_SUPPORTED_NAME
+      = "isThreadAllocatedMemorySupported";
+  private static final String SUN_THREAD_MXBEAN_IS_THREAD_ALLOCATED_MEMORY_ENABLED_NAME
+      = "isThreadAllocatedMemoryEnabled";
+  private static final String SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_NAME = "getThreadAllocatedBytes";
+  private static final Method SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_METHOD;
+
+  private static final ThreadMXBean MX_BEAN = ManagementFactory.getThreadMXBean();
+  private static final boolean IS_CURRENT_THREAD_CPU_TIME_SUPPORTED = MX_BEAN.isCurrentThreadCpuTimeSupported();
+  private static final boolean IS_THREAD_ALLOCATED_MEMORY_SUPPORTED;
+  private static final boolean IS_THREAD_ALLOCATED_MEMORY_ENABLED;
+  private static boolean _isThreadCpuTimeMeasurementEnabled = false;
+  private static boolean _isThreadMemoryMeasurementEnabled = false;
+  private final long _startTimeNs;
+  private final long _startBytesAllocated;
+
+  public ThreadResourceUsageProvider() {
+    _startTimeNs = _isThreadCpuTimeMeasurementEnabled ? MX_BEAN.getCurrentThreadCpuTime() : -1;
+
+    long startBytesAllocated1;
+    try {
+      startBytesAllocated1 = _isThreadMemoryMeasurementEnabled
+          ? (long) SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_METHOD.invoke(MX_BEAN, Thread.currentThread().getId()) : -1;
+    } catch (IllegalAccessException | InvocationTargetException e) {
+      startBytesAllocated1 = -1;
+      LOGGER.error("Exception happened during the invocation of getting initial bytes allocated", e);
+    }
+    _startBytesAllocated = startBytesAllocated1;
+  }
+
+  public static boolean isThreadCpuTimeMeasurementEnabled() {
+    return _isThreadCpuTimeMeasurementEnabled;
+  }
+
+  public static void setThreadCpuTimeMeasurementEnabled(boolean enable) {
+    _isThreadCpuTimeMeasurementEnabled = enable && IS_CURRENT_THREAD_CPU_TIME_SUPPORTED;
+  }
+
+  public static boolean isThreadMemoryMeasurementEnabled() {
+    return _isThreadMemoryMeasurementEnabled;
+  }
+
+  public static void setThreadMemoryMeasurementEnabled(boolean enable) {
+    _isThreadMemoryMeasurementEnabled = enable && IS_THREAD_ALLOCATED_MEMORY_SUPPORTED
+        && IS_THREAD_ALLOCATED_MEMORY_ENABLED;
+  }
+
+  public long getThreadTimeNs() {
+    return _isThreadCpuTimeMeasurementEnabled ? MX_BEAN.getCurrentThreadCpuTime() - _startTimeNs : 0;
+  }
+
+  public long getThreadAllocatedBytes() {
+    try {
+      return _isThreadMemoryMeasurementEnabled ? (long) SUN_THREAD_MXBEAN_GET_BYTES_ALLOCATED_METHOD
+          .invoke(MX_BEAN, Thread.currentThread().getId()) - _startBytesAllocated : 0;
```

Comment:
So if the memory mxbean is not supported, will the caller fallback on cpu time approach ?

Reply:
Yes, see `killMostExpensiveQuery`